README.md
<!-- Logo -->

<p align="center">
  <img width="500px" src="docs/Images/logo1x.png">
</p>

# Osu!StatQt

Osu statistics right on your desktop, made with Qt5

A lightweight desktop application to show you information about plays, scores and beatmaps. Basically a simple osu-web client.

## Status

![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/sortedcord/osu-statqt?include_prereleases&style=for-the-badge)
![GitHub last commit](https://img.shields.io/github/last-commit/sortedcord/osu-statqt?style=for-the-badge)
![GitHub issues](https://img.shields.io/github/issues/sortedcord/osu-statqt?label=ISSUES%20OPEN&style=for-the-badge)
[![wakatime](https://wakatime.com/badge/user/6141c7ad-332a-45eb-b78f-57b7202343fc/project/b0f1cad2-0279-4d71-9d75-1d434ecf82bb.svg?style=for-the-badge)](https://wakatime.com/badge/user/6141c7ad-332a-45eb-b78f-57b7202343fc/project/b0f1cad2-0279-4d71-9d75-1d434ecf82bb)


Currently, it is a passion project for me and I'm creating this to learn the pyQt framework and more stuff about Python. I plan on adding features to it with time and frequent releases as well until its fully complete or there are changes in osu API V2.

In its alpha stage, this is just the barebones with basic features. With the subsequent beta and stable releases, the featureset would only increase.

## Usage

![usage](docs/Images/usage0.png)

As of the current release, you can [set a default user](#setting-default-user) for osu stat and view your recent activity and scores (including failed scores) as well as their information such as Performance Points, Time, Beatmap Title, Diff and Ranked Score.

## Installing Osu!StatQt

**Latest Release:**

| [Windows 8.1+ (x64)](https://github.com/sortedcord/osu-statqt/releases) | [Linux (x64)](https://github.com/sortedcord/osu-statqt/releases)
| ------------- | ------------- |

Binaries are available for both Windows and Linux <i>(Needs more Testing)</i>. It's quite straightforward for windows as you simply have to run a `.exe` file. For Linux you may need to set the File as executable and then run it.

```bash
$ sudo chmod +x ./OsuStatQt-0.0.5-linux-x64-standalone

./OsuStatQt-0.0.5-linux-x64-standalone
```

## Configuring Osu!StatQt

<i>Note: Osu!StatQt generates a config file. Do not edit it manually by opening it in any text editor nor share it to anyone else!</i>

### Api Credentials

OsuStat uses Circleguard's [ossapiV2](https://github.com/circleguard/ossapi) which is a python wrapper for Osu-api V2. This requires you to create a OAuth Application in your [osu account settings](https://osu.ppy.sh/home/account/edit)

<i>Note: You can use an existing OAth Application's credentials if you have created one before.</i>

<summary>
<details>
Open OsuStatQt and go to <b> Preferences > Settings </b> and then click on `Get Credentials` Button.


You would have to login in case you are not already logged in then go to the account setting page. There you would need to scroll all the way down to the OAuth Section and click on new OAuth Application.

![](docs/Images/img1.png)

In the popup, set the application name as anything you want and for the redirect url, set it as `https://127.0.0.1:727`.


![](docs/Images/img2.png)

<i>You may need to change `727` to something else if you already have some process using that port.</i>

Entering the details will give you <b>Client ID</b> and <b>Client Secret</b>. Copy them to the OsuStatQt settings page (Make sure there are no spaces in either of the fields) and click `Verify`.

![](docs/Images/3.png)

If you entered the credentials properly, then you will get a message telling you so. 

![](docs/Images/img4.png)

</details>
</summary>

Once this is done, you can move on to setting up Default user and start using OsuStatQt.
<br>

### Setting Default User

In order to set a default user (which generally is you but it can be any user on Bancho), open settings and enter the default username in the user field. You have to enter the correct spelling else it may set it as somebody else or may not find the user at all!

## Building Osu!StatQt

I use `pyinstaller` for building and managing packages for OsuStatQt. It is highly recommended to use a virtual environment when building with pyinstaller.

[Circleguard's ossapi](https://github.com/circleguard/ossapi) causes issues with pyinstaller as for some reason its not able to read the dumped token files. Instead I suggest using https://github.com/sortedcord/ossapi which I have modified to not dump the tokens at all.


``` bash
# Assuming this is in a venv

pip install -e git+https://github.com/sortedcord/ossapi.git@46a3c0a684aa792f2f3f55001539cb4faf3fd8ba#egg=ossapi
```

Other dependencies which you will need include the following.

``` bash
pip install -r requirements.txt
```

And Clone this repository

``` bash

git clone https://github.com/sortedcord/osu-statqt.git
cd osu-statqt
```

Build files are included with in the source code. Use [onefile.spec](onefile.spec) to build a single file binary and use [build_package.spec](build_package.spec). Usage is like so:

```bash
$ pyinstaller ./build_package.spec #Or change this to onefile according to your need.
```

### Enable verbose (console)

Change `console=False` to `console=True` in `.spec` files.

### Mac OS (Catalina and above)

I don't own a Mac nor am I interested in setting up a hackintosh, so practically there is no way for me to test the mac os build, however the process of building binaries is more or less the same, the only catch is to build a standalone binary.

```bash
$ pyinstaller onefile.spec
```

## Credits

Would like to thank these creators for their projects and contributions.

- [ppy, osu-web API & osu-resources](https://github.com/peppy)
- [circleguard: osuAPI, licensed under GPL-3.0 License](https://github.com/circleguard/ossapi)


CHANGELOG.md
# Changelog

All notable changes to OsuStatQt will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.5] Alpha - 2022-01-09

### Added / Changes
- Cleaned Up and Optimized `settings.py`. The code is wayy more easy to read now. #21 
- Created a Top Bar with a Save Button to save all field changes . #23 
- Added Reset Button to delete existing config and reset all settings.
- Using Scrollbars in settings. #17 
- Implemented Logging #10 
- Added feature to Toggle Failed Scores #20 
- Added setting to change the number of items loading in panels on first load. #19 

### Bug Fixes
- Fixed Tabs Popping Out on Refreshing #25 
- Fixed Icons and images not showing up in Linux and MacOS #27 
- Included custom fonts in the assets #26 
- Refresh Button now reloads panels only once. #28 
- OsuStatQt now loads assets when executed from any location. 

## [0.0.4] Alpha - 2022-01-02

### Added / Changes
- Overhauled recent activity and scores pane.
- Stylized Scrollbars
- Panes now show rankscore images
- Other statistics such as beatmap title, diff, relative time is also shown
- The panes now have a "Show more" for pagination.

### Bug Fixes
- Fixed AttributeError: attribute UserSupportEvent
## [0.0.3] Alpha - 2021-12-30

### Added / Changes
- Added a cooldown to refresh button to prevent spamming
- Switched to pickling of config instead of dumping plain text.
- `Get Credentials` Button now opens the account settings.
- Achievement Names show up in Recent Activities

### Bug Fixes
- App crashing when the config file is blank or with missing values

## [0.0.2] Alpha - 2021-12-28

### Added / Changes
- Refresh Button can now be enabled from the settings automatically
- Recent activity and scores work now!

### Bug Fixes
- Could not set default user from settings UI
- Application crash when credentials were correct
- Default User Exception
- 2 Recent Scores Panel Showing up on refresh
- User Data Actually Shows Up now!


## [0.0.1] Alpha - 2021-12-27


### Added
- First release
- Created .osustat config wrapper
- API Credentials Authorization
- Shows Recent Activity and Recent Scores Beatmap Titles for User set as default.



[0.0.5]: https://github.com/sortedcord/osu-statqt/releases/tag/v0.0.5-alpha
[0.0.4]: https://github.com/sortedcord/osu-statqt/releases/tag/v0.0.4-alpha
[0.0.3]: https://github.com/sortedcord/osu-statqt/releases/tag/v0.0.3-alpha
[0.0.2]: https://github.com/sortedcord/osu-statqt/releases/tag/v0.0.2-alpha
[0.0.1]: https://github.com/sortedcord/osu-statqt/releases/tag/v0.0.1-alpha


settings.py
from PyQt5 import QtCore
from PyQt5.QtWidgets import *

from Components.settings import *
from Components.css_files import scrollbar_style
from Components.utility import CustomHLayout, CustomVLayout, MsgBox

from config import Config, del_config_file
# from functions import OsuStatUser
import webbrowser
from loguru import logger

class SettingsWindow(QMainWindow):
    def __init__(self, mainWindow):
        super().__init__()

        self.mainWindow = mainWindow
        self.refresh_cooldown_values = [0, 5000, 10000, 15000, 30000, 60000] # time in milliseconds

        self.setupUi()
        logger.debug("Loaded Window UI Structure")
        self.setupText()
        self.setupStylesheet()
        logger.debug("Applied Stylesheet to Qt")
        self.load_data()
        logger.debug("Loaded Window Data")


    # <=============== Clicked Events ===============>
    def verify_credentials_clicked(self):
        logger.info("Executing Verify Credentials Button clicked event")

        mainWindow = self.mainWindow
        mainWindow.config.client_id = self.client_id_field.text()
        mainWindow.config.client_secret = self.client_secret_field.text()
        logger.debug(f"Verifying {mainWindow.config.client_id},{mainWindow.config.client_secret} ")
        mainWindow.config.verify_credentials()

        # If credentials are correct
        if mainWindow.config.cred_verification_status == 'VERIFIED':
            logger.success("Credentials have been verified.")

            # Disable Text Fields
            logger.debug("Locked Text Fields")
            self.client_id_field.setEnabled(False)
            self.client_secret_field.setEnabled(False)

            # Hide verify Button
            logger.debug("Set Credentials button visibility to False")
            self.verify_credentials_button.setVisible(False)

            logger.debug("Remove Alert Frame")
            mainWindow.alert_frame.setParent(None)

            MsgBox("Authenticated Successfully, you may use this client now.", 'information')

            # Refresh Settings Window
            logger.debug("Remove Alert Frame")
            self.load_data()

        # If invalid credentials
        else:
            logger.error("Entered credentials are invalid")
            mainWindow.config.cred_verification_status = 'INVALID'
            logger.info(f"Updated Login credentials to {mainWindow.config.cred_verification_status}")
            MsgBox("Invalid Credentials. Please Try Again.", 'critical')
        
        # Refresh Settings Window
        self.load_data()
        logger.error("Reloaded settings Window fields")


    def get_credentials_clicked(self):
        logger.info("Get Credentials button click event performed")
        webbrowser.open("https://osu.ppy.sh/home/account/edit")
        logger.info("Opening Link in Browser")


    def save_config_clicked(self):
        logger.info("Save Config Button Click Event Triggered")

        mainWindow = self.mainWindow

        if mainWindow.config.cred_verification_status == 'VERIFIED':

            if mainWindow.config.default_user is None:
                logger.debug("Change in default user detected")
                self.save_default_user()
            else:
                if mainWindow.config.default_user.username != self.default_user_field.text():
                    logger.debug("Change in default user detected")
                    self.save_default_user()
            
            if mainWindow.config.refresh_cooldown != self.refresh_cooldown_values[self.refresh_limit_combo.currentIndex()]:
                logger.debug("Change in refresh cooldown  detected")
                self.save_refresh_cooldown()
            
            if mainWindow.config.show_failed_scores != self.toggle_failed_scores_checkbox.isChecked():
                logger.debug("Change in failed scores option")
                self.toggled_failed_scores()
                logger.info(f"Set Failed Scores to {mainWindow.config.show_failed_scores} in config")
            
            if mainWindow.config.panel_items != int(self.panel_items_field.text()) and self.panel_items_field.text() != '':
                logger.debug("Change in number of items shown on load.")
                self.save_panel_items()
                logger.info(f"Set panel items shown on load to {mainWindow.config.panel_items} in config")
            
            MsgBox("Settings Saved successfully.")

            mainWindow.config.dump_config()

        
        else:
            MsgBox("Please verify API credentials first.", "critical")


    def reset_config_clicked(self):
        mainWindow = self.mainWindow

        # Set MainWindow to new config instance
        mainWindow.config = Config()
        logger.info("Reset config instance")

        # Delete existing config file
        del_config_file()
        logger.info("Deleted Config File")

        # Reset Fields
        self.client_id_field.setText(None)
        self.client_secret_field.setText(None)
        self.default_user_field.setText(None)
        self.refresh_limit_combo.setCurrentIndex(3)
        logger.debug("Fields have been reset")

        # Load settings again
        self.load_data()
        


    def toggled_failed_scores(self):
        mainWindow = self.mainWindow

        mainWindow.config.show_failed_scores = self.toggle_failed_scores_checkbox.isChecked()


    # <=============== Save Events ===============>
    def save_default_user(self):
        mainWindow = self.mainWindow

        try:
            compact_user = mainWindow.config.api.search(query=self.default_user_field.text()).users.data[0]
        except:
            logger.error(f"{self.default_user_field.text()} was not found in Bancho")
            MsgBox('Specified user not found. Try checking your spelling.')
        else:  # If user is found
            logger.success("Specified User found.")
            mainWindow.config.default_user = compact_user
            logger.debug("Set Compact User")

            mainWindow.default_user_expand = compact_user.expand()
            MsgBox(f"Default user has been set as {mainWindow.config.default_user.username}", "information")

            self.default_user_field.setText(mainWindow.config.default_user.username)

            mainWindow.enable_refresh_button()
    

    def save_refresh_cooldown(self):
        if self.refresh_limit_combo.currentIndex() < 2:
            MsgBox("""Although you can, but it is highly recommended not to set the refresh cooldown value below 10 seconds as it may cause spamming of the refresh button.""", 'information')
        self.mainWindow.config.refresh_cooldown = self.refresh_cooldown_values[self.refresh_limit_combo.currentIndex()]


    def save_panel_items(self):
        if int(self.panel_items_field.text()) > 27:
            MsgBox("Do not set this value value too high as it may register as abuse of the API. Recommended value is in the range of 15 to 25.", "warning")
        self.mainWindow.config.panel_items = int(self.panel_items_field.text())

    # <=============== Other Functions ===============>
    def ulock_panels(self, state):
        self.default_user_panel.setEnabled(state)
        self.osuStat_panel.setEnabled(state)


    def load_data(self):
        mainWindow= self.mainWindow

        # If credentials verified
        if mainWindow.config.cred_verification_status == 'VERIFIED':
            self.ulock_panels(True)

            self.client_id_field.setText(str(mainWindow.config.client_id))
            self.client_secret_field.setText(mainWindow.config.client_secret)
            self.toggle_failed_scores_checkbox.setChecked(mainWindow.config.show_failed_scores)
            self.panel_items_field.setText(str(mainWindow.config.panel_items))
            logger.debug("Filled in Fields")

            # Disable Credentials Field
            self.client_id_field.setEnabled(False)
            self.client_secret_field.setEnabled(False)
            logger.debug("Locked Credentials Fields")

            # Hide verify credentials button
            self.verify_credentials_button.setVisible(False)
            logger.debug(f"Verify Credentials Button Visibility > {self.verify_credentials_button.isVisible}")

            # Load Default User
            if mainWindow.config.default_user is not None:
                self.default_user_field.setText(mainWindow.config.default_user.username)
                logger.debug(f"Default User Field set to {mainWindow.config.default_user.username}")
            else:
                logger.debug(f"Disabled Refresh Button on MainWindow.")
                mainWindow.disable_refresh_button()

        # If credentials not verified
        else:
            logger.debug("Credentials Not Verified. Locking Other Panels")
            self.ulock_panels(False)
            logger.debug("Disabled Refresh Button")
            mainWindow.disable_refresh_button()
        
        # Load Refresh Cooldown
        try:
            self.refresh_limit_combo.setCurrentIndex(self.refresh_cooldown_values.index(mainWindow.config.refresh_cooldown))
        except IndexError:
            logger.error("That value of refresh cooldown is not supported")



    def setupUi(self):
        # Window Layout
        self.resize(800, 439)
        self.centralwidget = QWidget(self)
        self.setCentralWidget(self.centralwidget)
        self.window_layout = CustomVLayout(self.centralwidget)



        # Top Title Bar
        self.top_bar_frame = QFrame(self.centralwidget)
        self.top_bar_frame.setMinimumSize(QtCore.QSize(800, 49))
        self.top_bar_frame.setMaximumSize(QtCore.QSize(16777215, 50))

        # Title Bar Layout
        self.title_bar_layout = CustomHLayout(self.top_bar_frame, (6,6,6,6), 12)
        self.window_layout.addWidget(self.top_bar_frame)

        # Title Bar Content
        self.title_label = QLabel(self.top_bar_frame)
        self.title_bar_layout.addWidget(self.title_label)

        # Reset Config Button
        self.reset_config_button = SettingsButton("Reset Config", "rgb(204,51,51)", "rgb(234,71,70)")
        self.reset_config_button.clicked.connect(lambda: self.reset_config_clicked())
        self.title_bar_layout.addWidget(self.reset_config_button)

        # Save Config Button
        self.save_config_button = SettingsButton("Save Config")
        self.save_config_button.clicked.connect(lambda: self.save_config_clicked())
        self.title_bar_layout.addWidget(self.save_config_button)

        # Settings Scrollable Area
        self.settings_scrollable_area = QScrollArea(self.centralwidget)
        self.settings_scrollable_area.setWidgetResizable(True)
        self.settings_scrollable_area.setStyleSheet(scrollbar_style)
        self.window_layout.addWidget(self.settings_scrollable_area)


        # Settings Content
        self.settings_content_frame = QFrame()
        self.settings_content_frame.setStyleSheet("border:none;")
        self.settings_scrollable_area.setWidget(self.settings_content_frame)

        # Settings Content Layout
        self.settings_layout = CustomVLayout(self.settings_content_frame, spacing=5)

        # Credentials Panel
        self.credentials_panel = SettingsPanel('Credentials')
        
        # Client ID
        self.client_id_label = SettingsLabel('client ID')
        self.client_id_label.setAlignment(QtCore.Qt.AlignRight | QtCore.Qt.AlignTrailing | QtCore.Qt.AlignVCenter)
        self.credentials_panel.form_frame_layout.setWidget(0, QFormLayout.LabelRole, self.client_id_label)

        self.client_id_field = SettingsField()
        self.credentials_panel.form_frame_layout.setWidget(0, QFormLayout.FieldRole, self.client_id_field)

        # Client Secret
        self.client_secret_label = SettingsLabel('client secret')
        self.credentials_panel.form_frame_layout.setWidget(1, QFormLayout.LabelRole, self.client_secret_label)

        self.client_secret_field = SettingsField()
        self.credentials_panel.form_frame_layout.setWidget(1, QFormLayout.FieldRole, self.client_secret_field)

        # Submit Credentials Button
        self.verify_credentials_button = SettingsButton('Verify')
        self.credentials_panel.form_frame_layout.setWidget(6, QFormLayout.FieldRole, self.verify_credentials_button)
        self.verify_credentials_button.clicked.connect(lambda: self.verify_credentials_clicked())

        # Get Credentials Button
        self.get_credentials_button = SettingsButton('Get Credentials')
        self.credentials_panel.form_frame_layout.setWidget(7, QFormLayout.FieldRole, self.get_credentials_button)
        self.get_credentials_button.clicked.connect(lambda: self.get_credentials_clicked())


        self.settings_layout.addWidget(self.credentials_panel)


        # Default User Panel
        self.default_user_panel = SettingsPanel('User')

        # Set Default User Label
        self.default_user_label = SettingsLabel('user')
        self.default_user_label.setAlignment(QtCore.Qt.AlignRight | QtCore.Qt.AlignTrailing | QtCore.Qt.AlignVCenter)   
        self.default_user_panel.form_frame_layout.setWidget(0, QFormLayout.LabelRole, self.default_user_label)

        self.default_user_field = SettingsField()
        self.default_user_panel.form_frame_layout.setWidget(0, QFormLayout.FieldRole, self.default_user_field)


        self.settings_layout.addWidget(self.default_user_panel)


        # OsuStat Panel
        self.osuStat_panel = SettingsPanel('OsuStat')
        
        # Refresh Cooldown
        self.refresh_cooldown_label = SettingsLabel('refresh cooldown')
        self.refresh_cooldown_label.setAlignment(QtCore.Qt.AlignRight|QtCore.Qt.AlignTrailing|QtCore.Qt.AlignVCenter)
        self.osuStat_panel.form_frame_layout.setWidget(0, QFormLayout.LabelRole, self.refresh_cooldown_label)

        self.refresh_limit_combo = QComboBox()
        self.refresh_limit_combo.addItem("")
        self.refresh_limit_combo.addItem("")
        self.refresh_limit_combo.addItem("")
        self.refresh_limit_combo.addItem("")
        self.refresh_limit_combo.addItem("")
        self.refresh_limit_combo.addItem("")
        self.osuStat_panel.form_frame_layout.setWidget(0, QFormLayout.FieldRole, self.refresh_limit_combo)

        # Toggle Failed Scores
        self.toggle_failed_scores = SettingsLabel('show failed scores')
        self.toggle_failed_scores.setAlignment(QtCore.Qt.AlignRight|QtCore.Qt.AlignTrailing|QtCore.Qt.AlignVCenter)
        self.osuStat_panel.form_frame_layout.setWidget(1, QFormLayout.LabelRole, self.toggle_failed_scores)

        # Toggle Failed Scores check box
        self.toggle_failed_scores_checkbox = QCheckBox()
        self.toggle_failed_scores_checkbox.setIconSize(QtCore.QSize(42, 16))
        sizePolicy = QSizePolicy(QSizePolicy.Minimum, QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.toggle_failed_scores_checkbox.sizePolicy().hasHeightForWidth())
        self.toggle_failed_scores_checkbox.setSizePolicy(sizePolicy)
        self.toggle_failed_scores_checkbox.setText("")
        self.osuStat_panel.form_frame_layout.setWidget(1, QFormLayout.FieldRole, self.toggle_failed_scores_checkbox)

        # Panel Items Label
        self.panel_items_label = SettingsLabel('Items shown on load')
        self.panel_items_label.setAlignment(QtCore.Qt.AlignRight|QtCore.Qt.AlignTrailing|QtCore.Qt.AlignVCenter)
        self.osuStat_panel.form_frame_layout.setWidget(2, QFormLayout.LabelRole, self.panel_items_label)

        # Panel Items Field 
        self.panel_items_field = SettingsField()
        self.osuStat_panel.form_frame_layout.setWidget(2, QFormLayout.FieldRole, self.panel_items_field)


        self.settings_layout.addWidget(self.osuStat_panel)

        

    def setupText(self):
        self.setWindowTitle("Settings | OsuStatQt")
        self.title_label.setText("      settings")
        self.refresh_limit_combo.setItemText(0, "No Limit")
        self.refresh_limit_combo.setItemText(1, "Every 5 Seconds")
        self.refresh_limit_combo.setItemText(2, "Every 10 Seconds")
        self.refresh_limit_combo.setItemText(3, "Every 15 Seconds")
        self.refresh_limit_combo.setItemText(4, "Every 30 Seconds")
        self.refresh_limit_combo.setItemText(5, "Every Minute")
    


    def setupStylesheet(self):
        self.setStyleSheet("""
            background-color:rgb(24,22,29);
            color: rgb(255,255,255);
            padding:0px;
        """)

        self.top_bar_frame.setStyleSheet("""
            background-color: rgb(35,31,47);
        """)

        self.title_label.setStyleSheet("font: 63 18pt \"Torus Pro SemiBold\";")    

        self.refresh_limit_combo.setStyleSheet("""
            background-color: rgb(61, 57, 70);
            border: none;
            padding: 6px;
            border-radius: 4px;
            color: rgb(255,255,255);
            font: 63 10pt \"Torus Pro SemiBold\";
        """)

        self.toggle_failed_scores_checkbox.setStyleSheet("""
            background-color:rgb(49,47,56);
        """)
                                     

